runtime/src/iree/base/tree.c

```diff
@@ -0,0 +1,549 @@
+                    bool (*callback)(iree_tree_node_t*, void*),
+                    void* user_data) {
+  iree_tree_walk_helper(tree->root, walk_type, callback, user_data);
+}
+
+iree_tree_node_t* iree_tree_node_next(iree_tree_node_t* node) {
+  IREE_ASSERT(node != NULL);
+  // 1. to find the smallest thing on our right hand side.
+  if (!node->right->is_sentinel) {
+    node = node->right;
+    while (!node->left->is_sentinel) {
+      node = node->left;
+    }
+    return node;
+  }
+
+  // 2. Find the parent who is not on the right
+  iree_tree_node_t* parent = node->parent;
+  while (parent && node == parent->right) {
+    node = parent;
+    parent = node->parent;
+  }
+  return parent;
+}
+
+iree_tree_node_t* iree_tree_node_prev(iree_tree_node_t* node) {
+  IREE_ASSERT(node != NULL);
+  // 1. to find the largest thing on our left hand side.
+  if (!node->left->is_sentinel) {
+    node = node->left;
+    while (!node->right->is_sentinel) {
+      node = node->right;
+    }
+    return node;
+  }
+
+  // 2. Find the parent who is not on the left
+  iree_tree_node_t* parent = node->parent;
+  while (parent && node == parent->left) {
+    node = parent;
+    parent = node->parent;
+  }
+  return parent;
+}
+
+void* iree_tree_node_get_data(iree_tree_node_t* node) {
+  IREE_ASSERT(node);
+  return node->data;
+}
+
+iree_host_size_t iree_tree_node_get_key(iree_tree_node_t* node) {
+  IREE_ASSERT(node);
+  return node->key;
+}
+
+iree_tree_node_t* iree_tree_lower_bound(iree_tree_t* tree,
+                                        iree_host_size_t key) {
+  iree_tree_node_t* node = tree->root;
+  iree_tree_node_t* last = NULL;
+  while (node->is_sentinel == false) {
+    last = node;
+    if (key == node->key) {
+      return node;
+    } else if (key < node->key) {
+      node = node->left;
+    } else {
+      node = node->right;
+    }
+  }
+  if (!last || last->key > key) {
+    return last;
+  }
+  return iree_tree_node_next(last);
+}
+
+iree_tree_node_t* iree_tree_upper_bound(iree_tree_t* tree,
+                                        iree_host_size_t key) {
+  iree_tree_node_t* node = tree->root;
+  iree_tree_node_t* last = NULL;
+  while (node->is_sentinel == false) {
+    last = node;
+    if (key == node->key) {
+      return node;
+    } else if (key < node->key) {
+      node = node->left;
+    } else {
+      node = node->right;
+    }
+  }
+  if (!last || last->key > key) {
+    return last;
+  }
+  while (last && last->key <= key) {
+    last = iree_tree_node_next(last);
+  }
+  return last;
+}
+
+static void iree_tree_replace(iree_tree_t* tree, iree_tree_node_t* dst,
+                              iree_tree_node_t* src) {
+  if (!dst->parent) {
+    tree->root = src;
+  } else if (dst == dst->parent->left) {
+    dst->parent->left = src;
+  } else {
+    dst->parent->right = src;
+  }
+  src->parent = dst->parent;
+}
+
+static void iree_tree_remove(iree_tree_t* tree, iree_tree_node_t* z) {
+  iree_tree_node_t *x, *y = NULL;
```

Comment:
initialize on their own lines
```suggestion
  iree_tree_node_t* x = NULL;
  iree_tree_node_t* y = z;
```

I don't like the variable naming here - were this longer lived it'd be good to name these so someone can read it - only time to use names like this is if you're referencing a paper (that is linked on the type documentation) so that the algorithm can be matched up.